src/agent/block_store_services/block_store_fs.js

```diff
@@ -249,6 +249,60 @@ class BlockStoreFs extends BlockStoreBase {
         }
     }
 
+
+    /**
+     * @note Tier2 API only
+     * @param {string[]} block_ids
+     * @returns {Promise<{ evicted_block_ids: string[] }>}
+     */
+    async _evict_blocks(block_ids) {
+        const evicted_block_ids = [];
+        await P.map_with_concurrency(10, block_ids, async block_id => {
+            try {
+                if (await this._evict_block(block_id)) evicted_block_ids.push(block_id);
+            } catch (err) {
+                dbg.warn(`evict block ${block_id} failed due to`, err);
+            }
+        });
+
+        return { evicted_block_ids };
+    }
+
+    async _evict_block(block_id) {
+        const fs_context = this.fs_context;
+        const block_path = this._get_block_data_path(block_id);
+        let evicting = false;
+        let file = null;
+
+        if (config.BLOCK_STORE_FS_TIER2_ENABLED) {
+            try {
+                file = await nb_native().fs.open(fs_context, block_path);
+                const stat = await file.stat(fs_context, { xattr_get_keys: [config.BLOCK_STORE_FS_XATTR_QUERY_MIGSTAT] });
+                const migstat = JSON.parse(stat?.xattr?.[config.BLOCK_STORE_FS_XATTR_QUERY_MIGSTAT] || '{}');
+
+                // If the block is in PREMIGRATED state, we need to MIGRATE it back to the
+                // tape storage (eviction).
+                if (migstat.State === 'PREMIGRATED') {
```

Comment:
@guymguym @piste-jp-ibm when is a file marked `RESIDENT`?

I had the following state transition assumptions:

`* - (write)-> MIGRATED - (recall) -> PREMIGRATED - (evict) -> MIGRATED`

Reply:
That's right @utkarsh-pro AFAIU `RESIDENT` will only be file files were written but never been migrated. Currently we always migrate blocks right after write+fsync, so that's going to be needed ATM. I do think that deferring the migration of freshly written blocks to the worker to batch them might be a desirable option to use disks as a buffer for the archive tier, and in that case we would change to avoid migrating on write, and then need the worker to also handle `RESIDENT` blocks. But we can change it later too, I just noticed it now.